src/drivers/src/EncoderOdometryNode.cpp

```diff
@@ -0,0 +1,290 @@
+    double wheel_circumference = 2 * M_PI * wheel_radius;
+    double d_l = left_wheel_ticks * wheel_circumference / ticks_per_rotation;
+    double d_r = right_wheel_ticks * wheel_circumference / ticks_per_rotation;
+
+    // TODO: Detailed doc comments here
+    // let p_t be the change from the previous state (x,y,theta): p_t = [x_t; y_t; theta_t]
+
+    // Cov[p_t] = E[p_t * transpose(p_t)] - E[p_t] * E[transpose(p_t)]
+
+    // Compute E[p_t * transpose(p_t)] - E[p_t]
+    double k_1 = std::pow(d_r, 2) + std::pow(d_l, 2);
+    double k_2 = std::pow(d_r, 2) - std::pow(d_l, 2);
+    double k_3 = d_r * d_l;
+    double k_4 = left_encoder_variance + right_encoder_variance;
+    double k_5 = left_encoder_variance - right_encoder_variance;
+
+    // let c_ij be the element at (i,j) in E[p_t * transpose(p_t)] - E[p_t]
+    double c_11 = (k_1 + 2*k_3 + k_4)/4 * cos(prev_yaw) * cos(prev_yaw);
+    double c_12 = (k_1 + 2*k_3 + k_4)/4 * cos(prev_yaw) * sin(prev_yaw);
+    double c_13 = (k_2 + k_5)/(2 * wheelbase_length) * cos(prev_yaw);
+    double c_21 = c_12;
+    double c_22 = (k_1 + 2*k_3 + k_4)/4 * sin(prev_yaw) * sin(prev_yaw);
+    double c_23 = (k_2 + k_5)/(2 * wheelbase_length) * sin(prev_yaw);
+    double c_31 = c_13;
+    double c_32 = c_23;
+    double c_33 = (k_2 + k_5)/(2 * wheelbase_length) * sin(prev_yaw);
+
+    // Compute E[p_t] * E[transpose(p_t)]
+    double k_6 = (d_r + d_l)/2 * cos(prev_yaw);
+    double k_7 = (d_r + d_l)/2 * sin(prev_yaw);
+    double k_8 = (d_r - d_l)/wheelbase_length;
+
+    // let d_ij be the element at (i,j) in E[p_t] * E[transpose(p_t)]
+    double d_11 = k_6*k_6;
+    double d_12 = k_6*k_7;
+    double d_13 = k_6*k_8;
+    double d_21 = d_12;
+    double d_22 = k_7*k_7;
+    double d_23 = k_7*k_8;
+    double d_31 = d_13;
+    double d_32 = d_23;
+    double d_33 = k_8 * k_8;
+
+    // Cov[p_t] = E[p_t * transpose(p_t)] - E[p_t] * E[transpose(p_t)]
+    // Let cov_A_B be the covariance between A and B
+    double cov_x_x = c_11 - d_11;
+    double cov_y_y = c_22 - d_22;
+    double cov_theta_theta = c_33 - d_33;
+    double cov_x_y = c_12 - d_12;
+    double cov_y_x = c_21 - d_21;
+    double cov_x_theta = c_13 - d_13;
+    double cov_theta_x = c_31 - d_31;
+    double cov_y_theta = c_23 - d_23;
+    double cov_theta_y = c_32 - d_32;
+
+    // Set the covariances at the correct places in the array
+    // (since the array also includes linear z and angular x and y)
+    geometry_msgs::PoseWithCovariance::_covariance_type covariances;
+    covariances.fill(0);
+    covariances[0*6 + 0] = cov_x_x;
+    covariances[0*6 + 1] = cov_x_y;
+    covariances[0*6 + 5] = cov_x_theta;
+    covariances[1*6 + 0] = cov_y_x;
+    covariances[1*6 + 1] = cov_y_y;
+    covariances[1*6 + 2] = cov_y_theta;
+    covariances[5*6 + 0] = cov_theta_x;
+    covariances[5*6 + 1] = cov_theta_y;
+    covariances[5*6 + 5] = cov_theta_theta;
+
+    return covariances;
```

Comment:
Further inspection, I think that this node is overlapping with the functionality of the EKF node. Awks. The paper you found was used to compare against EKF in performance. Though in comparison, it seems that due to the linearization in the EKF, it is not good at adapting angle changes that are large. Imagine inverse tangential plot. So in theory this filter should be better. But it does not take in to account other sensors. Perhaps the best way to approach this problem is to merge the covariance calculations. Off the top of my head, our covariances for this calculation is convoluted while the covariances with EKF will be linear, because assumed independent events. 

TL;DR - This is overlapped functionality with EKF. I don't think we should integrate the covariance calculations here into the EKF, but figure out how to use it on its own. I will need to think some more about this. Sorry.

Reply:
Clarifications: 

- The paper you use also assume independence between sensors.

Reply:
We can probably try to be cheeky and incorporate their pose estimation into our EKF and modify the covariance estimations to be nonlinear as well.

Reply:
Darn. Ok, we may have to nuke much of this anyways. Val and I did some testing, and it looks like `robot_pose_ekf` (which it looks like we might be using unless the custom EKF gets done soon) doesn't work if there are any numbers close to 0 in the covariance :cry:

Reply:
Basically what happens when you don't have covariance is that they would be considered independent events. Therefore mathematically speaking, when we perform the update step in our EKF our Kalman Gain will be Minimized and the updated data will be ignored.

The other way to think about it is that, since they're independent events, Kalman Filter will basically scrap the data that round and not use the data to make predictions that round.

Reply:
Hmm, ok, but it doesn't publish an update to the state at all (I suppose because it doesn't have any reliable odometry data?)